### PR TITLE
IBM MQ Metricbeat module

### PR DESCRIPTION
**Pull request by felix-lessoer**
This is the metricbeat module for IBM MQ. Which will be the transformation of [this standalone beat](https://github.com/felix-lessoer/qbeat)

@ruflin is helping to integrate this as this is the first x-pack module for metricbeat

### CODE REVIEW

**Comment by felix-lessoer** [comment]
The module is now able to collect the MQ metrics. 
A few things are open:

1.) In the module each metricseat is collected in parallel. This is not possible with the MQ lib at the moment. @ruflin Is there a way to collect the metricsets sequentially without putting everything in one fetch metricset?
If not I need to change the way I connect to MQ. Each metricset needs to connect to its own command / reply queue . 

2.) To build metricbeat with ibmmq module it needs to have the MQ C library available locally under 

for Linux
* cgo !windows CFLAGS: -I/opt/mqm/inc -D_REENTRANT
* cgo !windows LDFLAGS: -L/opt/mqm/lib64 -lmqm_r -Wl,-rpath=/opt/mqm/lib64 -Wl,-rpath=/usr/lib64
for Windows
* cgo windows CFLAGS:  -I"C:/Program Files/IBM/MQ/Tools/c/include"
* cgo windows LDFLAGS: -L "C:/Program Files/IBM/MQ/bin64" -lmqm

This needs to be switched to dynamic linking in mqi.go

3.) The PubSub metricset needs to be added

4.) The imports are pointing to my repo. This needs to be switched when it will be merged into Elastic repo

**Comment by ruflin** [comment]
That all metricset fetches are made in parallel is by design. To run it sequentially it would require quite fundamental changes. I would definitively recommend that each metricset maintains it's own connection, so if one metricset is blocked it does not impact others. Are there some advantages to this?

**Comment by ruflin** [comment]
Some notes on my end for this PR:

* We need to add a docker file for testing and generate the some expected outputs
* Perhaps split PR up per metricset
* Check what the impact of the dependencies on the size of the binary is
* Add libraries with govendor to make sure they are properly vendored and in the correct directory.

**Comment by ruflin** [comment]
Here is the docker image that can be used for testing: https://hub.docker.com/r/ibmcom/mq/

**Comment by felix-lessoer** [comment]
> Here is the docker image that can be used for testing: https://hub.docker.com/r/ibmcom/mq/

You need to create it with a Queue Manager Name
MQ_QMGR_NAME - Set this to the name you want your Queue Manager to be created with.

This Queue Manager Name needs to be in the configuration.
The connection config (cc) for Docker needs to look something like this to get a result:

  cc:
    clientMode: true
    mqServer: "DEV.ADMIN.SVRCONN/TCP/127.0.0.1 (1416)"
    userId: "admin"
    password: "passw0rd"

File: x-pack/metricbeat/module/ibmmq/_meta/config.reference.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+- module: ibmmq
+  metricsets: ["qmgr"]
```
Comment (ruflin) [inline x-pack/metricbeat/module/ibmmq/_meta/config.reference.yml]: We will need to complete this list.

File: x-pack/metricbeat/module/ibmmq/lib/discover.go [hunk 2]
```diff
@@ -0,0 +1,746 @@
+package ibmmqlib
```
Comment (ruflin) [inline x-pack/metricbeat/module/ibmmq/lib/discover.go]: This must be pulled in as vendor directory to make the separation clear.

File: x-pack/metricbeat/module/ibmmq/_meta/docs.asciidoc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+This is the IBM MQ module.
+
+The default metricsets are `qmgr`, `queue` and `channel`.
+You also can add any PCF command by using the 'advanced' metricset.
+
+[float]
+=== Compability
+
+This module is tested with IBM MQ v... TODO .
```
Comment (ruflin) [inline x-pack/metricbeat/module/ibmmq/_meta/docs.asciidoc]: Complete the TODO in the docs

**Comment by sorantis** [comment]
Resolves: https://github.com/elastic/beats/issues/13508

File: x-pack/metricbeat/module/ibmmq/channel/_meta/docs.asciidoc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+This is the IBM MQ module.
+
+The default metricsets are `qmgr`, `queue` and `channel`.
+You also can add any PCF command by using the 'advanced' metricset.
+
+[float]
+=== Compability
+
+This module is tested with IBM MQ v... TODO .
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/ibmmq/channel/_meta/docs.asciidoc]: Please keep in mind to complete this TODO, before merging.

File: x-pack/metricbeat/module/ibmmq/custom/_meta/docs.asciidoc [hunk 5]
```diff
@@ -0,0 +1,9 @@
+This is the IBM MQ module.
+
+The default metricsets are `qmgr`, `queue` and `channel`.
+You also can add any PCF command by using the 'advanced' metricset.
+
+[float]
+=== Compability
+
+This module is tested with IBM MQ v... TODO .
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/ibmmq/custom/_meta/docs.asciidoc]: See, https://github.com/elastic/beats/pull/8870/files#r325675853

File: x-pack/metricbeat/module/ibmmq/qmgr/_meta/docs.asciidoc [hunk 6]
```diff
@@ -0,0 +1,9 @@
+This is the IBM MQ module.
+
+The default metricsets are `qmgr`, `queue` and `channel`.
+You also can add any PCF command by using the 'advanced' metricset.
+
+[float]
+=== Compability
+
+This module is tested with IBM MQ v... TODO .
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/ibmmq/qmgr/_meta/docs.asciidoc]: same as https://github.com/elastic/beats/pull/8870/files#r325675853

File: x-pack/metricbeat/module/ibmmq/queue/_meta/docs.asciidoc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+This is the IBM MQ module.
+
+The default metricsets are `qmgr`, `queue` and `channel`.
+You also can add any PCF command by using the 'advanced' metricset.
+
+[float]
+=== Compability
+
+This module is tested with IBM MQ v... TODO .
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/ibmmq/queue/_meta/docs.asciidoc]: same as https://github.com/elastic/beats/pull/8870/files#r325675853

**Comment by ChrsMark** [comment]
> Some notes on my end for this PR:
> 
> * We need to add a docker file for testing and generate the some expected outputs
> * Perhaps split PR up per metricset
> * Check what the impact of the dependencies on the size of the binary is
> * Add libraries with govendor to make sure they are properly vendored and in the correct directory.

Hi @ruflin and @felix-lessoer!

To add to the above @ruflin's list, here are my very first comments/thoughts on this PR:

-  I will agree that we need to import `ibmmqlib` as external library
-  I would definitely encourage @felix-lessoer to split this PR into smaller ones. One for each metricset, so as to keep it track easier and review/merge things in a more standalone way. We did the same for https://github.com/elastic/beats/pull/9825 (see also https://github.com/elastic/beats/issues/10071) and was really helpful both for the reviewers and for the contributors.
- Then hound errors should be fixed.

I propose we start with these first, but let me know what you think!

**Comment by felix-lessoer** [comment]
@ChrsMark I made the initial split. Removed the biggest part of the ibmmqlib. The rest is the interface for every metricset to connect to IBM MQ. 
I also made a rebase.

File: x-pack/metricbeat/modules.d/ibmmq.yml.disabled [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Module: ibmmq
+# Docs: https://www.elastic.co/guide/en/beats/metricbeat/master/metricbeat-module-ibmmq.html
+
+- module: ibmmq
+  #metricsets:
+  #  - qmgr
+  #  - queue
```
Comment (ChrsMark) [inline x-pack/metricbeat/modules.d/ibmmq.yml.disabled]: These need to be removed from this PR I think. Will be added in each follow-up.

File: x-pack/metricbeat/include/list.go [hunk 9]
```diff
@@ -14,6 +14,11 @@ import (
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/s3_daily_storage"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/s3_request"
 	_ "github.com/elastic/beats/x-pack/metricbeat/module/aws/sqs"
+	_ "github.com/elastic/beats/x-pack/metricbeat/module/ibmmq"
+	_ "github.com/elastic/beats/x-pack/metricbeat/module/ibmmq/channel"
```
Comment (ChrsMark) [inline x-pack/metricbeat/include/list.go]: I think that only `qmgr` import should be here.

File: x-pack/metricbeat/module/ibmmq/_meta/config.reference.yml (hunk 1 above)
Comment (ChrsMark) [reply x-pack/metricbeat/module/ibmmq/_meta/config.reference.yml]: This is ok for now.

File: x-pack/metricbeat/module/ibmmq/_meta/config.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+# Module: ibmmq
+# Docs: https://www.elastic.co/guide/en/beats/metricbeat/master/metricbeat-module-ibmmq.html
+
+- module: ibmmq
+  #metricsets:
+  #  - qmgr
+  #  - queue
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/ibmmq/_meta/config.yml]: we need to keep only `qmgr`

File: x-pack/metricbeat/main.go [hunk 11]
```diff
@@ -15,6 +15,8 @@ import (
 	"os"
 
 	"github.com/elastic/beats/x-pack/metricbeat/cmd"
+	_ "github.com/felix-lessoer/beats/x-pack/metricbeat/module/ibmmq"
```
Comment (ChrsMark) [inline x-pack/metricbeat/main.go]: this should be under `github.com/elastic/beats/...`

**Comment by ChrsMark** [comment]
Currently in order to tackle the dependency to C library of ibm client we include the library files of total size `25M    ./module/ibmmq/ibmqlib/` .In addition we tune magefile so as to set the required env vars accordingly. Consequently in this state the PR is functional and the ibm module can work like this:
```console
mage build (in x-pack/metricbeat directory)
./metricbeat modules enable ibmmq
./metricbeat
```

* Searching if we could possibly remove some unwanted files of big size it seems that there is nothing big enough to remove.
** The included libraries should be checked if are compatible with multiple distros.

cc: @felix-lessoer @ruflin @sayden @odacremolbap

**Comment by ChrsMark** [comment]
Shifting to a different approach to use a dynamically loaded plugin to connect to IBM. This plugin is a thin wrapper on top of `mq-golang` client.

This wrapper is temporarily on https://github.com/ChrsMark/ibm_client_plugin.

## How to test on MacOS:
First download the IBM client library. This is a dynamically loaded library:
```
wget https://github.com/ChrsMark/ibm_client_plugin/raw/master/ibmqlib_darwin.so
```
Then set the IBM_LIBRARY_PATH variable to point to the downloaded file:
```
export IBM_LIBRARY_PATH=~/Downloads/ibmqlib_darwin.so
```
Then you are ready to go. You can build and execute Metricbeat and enable the module to collect metrics data
from IBM MQ:
```
go build  (or mage build)
./metricbeat modules enable ibmmq
./metricbeat -e -d "*"
```

Note that if we go with this approach we need to build and upload the plugins for Windows and linux that will use the respective Redistributable Clients accordingly.

**Comment by mtojek** [comment]
I believe we can close this PR in favor of a simple light module implementation.

**Pull request closed without merging**